File: csharp/P0001_TwoSum/Solutions/HashMapOnePass.cs
// Time complexity: O(n)
// Space complexity: O(n)

namespace P0001_TwoSum.Solutions.HashMapOnePass;

public class Solution
{
    public int[] TwoSum(int[] nums, int target)
    {
        Dictionary<int, int> seen = new Dictionary<int, int>();

        for (int i = 0; i < nums.Length; i++)
        {
            int complement = target - nums[i];

            if (seen.ContainsKey(complement))
                return [seen[complement], i];

            seen[nums[i]] = i; 
        }

        // This should be unreachable
        return [-1];
    }
}

File: rust/p0001_two_sum/src/hashmap_one_pass.rs
// Time complexity: O(n)
// Space complexity: O(n)

use std::collections::HashMap;

pub struct Solution;

impl Solution {
    pub fn two_sum(nums: Vec<i32>, target: i32) -> Vec<i32> {
        let mut seen: HashMap<i32, usize> = HashMap::new();

        for (i, num) in nums.iter().enumerate() {
            let complement = target - num;

            if let Some(&j) = seen.get(&complement) {
                return vec![j as i32, i as i32];
            }

            seen.insert(*num, i);
        }

        unreachable!("This should not be reachable with valid input");
    }
}
